src/deckgl-layers/layer-utils/cpu-aggregator.js

```diff
@@ -0,0 +1,606 @@
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+/* eslint-disable guard-for-in */
+import {experimental, AGGREGATION_OPERATION} from '@deck.gl/aggregation-layers';
+import {aggregate} from 'utils/aggregate-utils';
+import {AGGREGATION_TYPES, SCALE_FUNC} from 'constants/default-settings';
+
+const {BinSorter} = experimental;
+
+export const DECK_AGGREGATION_MAP = {
+  [AGGREGATION_OPERATION.SUM]: AGGREGATION_TYPES.sum,
+  [AGGREGATION_OPERATION.MEAN]: AGGREGATION_TYPES.average,
+  [AGGREGATION_OPERATION.MIN]: AGGREGATION_TYPES.minimum,
+  [AGGREGATION_OPERATION.MAX]: AGGREGATION_TYPES.maximum
+};
+
+export function getValueFunc(aggregation, accessor) {
+  const op =
+    AGGREGATION_OPERATION[aggregation.toUpperCase()] ||
+    AGGREGATION_OPERATION.SUM;
+  const keplerOp = DECK_AGGREGATION_MAP[op];
+
+  return pts => aggregate(pts.map(accessor), keplerOp);
+}
+
+export function getScaleFunctor(scaleType) {
+  return SCALE_FUNC[scaleType] || SCALE_FUNC.quantize;
+}
+
+function nop() {}
+
+export function getGetValue(step, props, dimensionUpdater) {
+  const {key} = dimensionUpdater;
+  const {value, weight, aggregation} = step.triggers;
+
+  let getValue = props[value.prop];
+
+  if (getValue === null) {
+    // If `getValue` is not provided from props, build it with aggregation and weight.
+    getValue = getValueFunc(props[aggregation.prop], props[weight.prop]);
+  }
+
+  if (getValue) {
+    this._setDimensionState(key, {getValue});
+  }
+}
+
+export function getDimensionSortedBins(step, props, dimensionUpdater) {
+  const {key} = dimensionUpdater;
+  const {getValue} = this.state.dimensions[key];
+
+  const sortedBins = new BinSorter(this.state.layerData.data || [], {
+    getValue,
+    filterData: props._filterData
+  });
+  this._setDimensionState(key, {sortedBins});
+}
+
+export function getDimensionValueDomain(step, props, dimensionUpdater) {
+  const {key} = dimensionUpdater;
+  const {
+    triggers: {lowerPercentile, upperPercentile, scaleType}
+  } = step;
+
+  if (!this.state.dimensions[key].sortedBins) {
```

Comment:
maybe use lodash get here to avoid any possible null values

Reply:
this is temporary until I can merge this into deck.gl. Deck.gl doesn't have lodash